common/configurable/src/main/java/io/helidon/common/configurable/ThreadPool.java

```diff
@@ -498,31 +505,76 @@ void setPool(ThreadPool pool) {
 
         @Override
         public boolean offer(Runnable task) {
-            return enqueue(task);
+            if (!semaphoreWrite.tryAcquire()) {
+               return false;
+            }
+            enqueue(task);
```

Comment:
Naïve reviewer comment: if this throws a (runtime) exception, `semaphoreRead` will remain unreleased, right?  Can `enqueue` throw an exception?

Reply:
I asked myself that question too. Since the new queue is really unbounded, only limited by the semaphores, maybe only an NPE can be thrown. However, perhaps using a finally block is more elegant.

Reply:
Please ignore latest commit, did not intent to push. Fixing now.

Reply:
A `finally` block can give you unwanted results, if indeed anything is thrown. Ie a `release` will result in reader/writer making progress, and polling `null` from the queue.

It makes sense to check the `task` is non-null (before acquiring `Semaphore`). Then we will rely on `j.u.c.ConcurrentLinkedQueue` not throwing.

Alternatively, in `enqueue` catch exceptions and release `semaw` (not `semar`) upon exceptions or if `super.offer` returns `false` (should never do that for `ConcurrentLinkedQueue`).

Reply:
eg 
```
private boolean enqueue(Runnable task) {
   boolean r=false;
   try {
             r = super.offer(task);
   } finally {
      if (!r) {
         semaw.release(); // let another offer/put try their luck
      }
   }
```

Reply:
I accidentally pushed those changes from Intellij. Please check final version.